### PR TITLE
Properly render in listView (showColumns=true) mode

### PR DESCRIPTION
**Pull request by amenk**
When showColumns is on, DcGeneral expects an array with the values to be returned.

Currently you have to set 
\ContaoCommunityAlliance\DcGeneral\DataDefinition\Definition\View\DefaultListingConfig::getShowColumns

to true to see an effect

### CODE REVIEW

**Comment by amenk** [comment]
showColumns can now be configured in input screen

File: src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php [hunk 1]
```diff
@@ -10,7 +10,6 @@
  * @package    MetaModels
  * @subpackage Core
  * @author     Christian Schiffler <c.schiffler@cyberspectrum.de>
- * @author     David Molineus <david.molineus@netzmacht.de>
```
Comment (discordier) [inline src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php]: Why did you remove him? He has contributed to this file.

File: src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php [hunk 2]
```diff
@@ -157,41 +167,6 @@ public static function getReadableValue(RenderReadablePropertyValueEvent $event)
     }
 
     /**
-     * Add additional parent header fields.
```
Comment (discordier) [inline src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php]: Are you really sure this listener is not needed anymore? This is the parent header for parented MetaModels and not the child list.

File: src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php (hunk 1 above)
Comment (amenk) [inline src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php]: War ein Versehen. Ich entwickle leider momentan nicht mit dem aktuellen git stand sondern im Kundenprojekt und hab die Datei rüber kopiert. Sorry.

File: src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php (hunk 2 above)
Comment (amenk) [inline src/MetaModels/DcGeneral/Events/MetaModel/RenderItem.php]: gleiches Problem wie oben. Ich arbeite jetzt mit `patch`.

**Comment by amenk** [comment]
Now the question is on how to hide the DCA properties id, pid, sorting, timestamp from the listview. Is it save to remove them from the PropertiesDefinition?

screenshot_selection_031

**Comment by discordier** [comment]
If you remove them from the definition, they will become unavailable throughout the system. Therefore DON'T.
You are here in the part that is building the data definition (most closely comparable with the container building in symfony) but you simply want to change the display output therefore you should stick to change the model rendering by attching to any of the various rendering events.

**Comment by amenk** [comment]
Looks like I have to make an additional change to DC_General then? Because \ContaoCommunityAlliance\DcGeneral\Contao\View\Contao2BackendView\ListView::getTableHead is directly accessing the properies... Do you have any advice?

**Comment by discordier** [comment]
@dmolineus is currently reworking the whole model and property rendering AFAIR, so there might come something in handy this way.
Currently I lack the time to look in detail into this, sorry.

**Comment by dmolineus** [comment]
@amenk Looks like the dcg should be fixed here. The DC_Table uses `list.label.fields` here. This should be done for the DCG as well.

**Hint:** Have a look on the legacy mapping to find the proper definition.
http://contao-community-alliance.github.io/dc-general-docs/docs/data_definition/container/legacy_mapping.html

I currently working on the DCG because I was put in charge of. Fixing the column layout is not part of it. 

In general I think there could be some improvements on the list implementation but the dcg is still a huge construction side. ;)

**Comment by amenk** [comment]
the doc mentiones list.label.fields -> -> (required) used for ListItemFormatter::$propertyNames
but there is no ListItemFormatter. I am using $listing->headerPropertyNames now. Hope that is okay.

**Comment by dmolineus** [comment]
Nope, it's not designed for. You'd better take the time to understand the concepts of the dcg better even when it's documentation if far away from being complete. ;)

**Hint:** https://github.com/contao-community-alliance/dc-general/blob/develop/src/ContaoCommunityAlliance/DcGeneral/Contao/Subscriber/FormatModelLabelSubscriber.php#L46

**Comment by amenk** [comment]
What is the "nope" for? Looks like now I have to transfer the header properties to $formatter->propertyNames ?

**Comment by dmolineus** [comment]
> What is the "nope" for? 

It was meant for:

> I am using $listing->headerPropertyNames now. Hope that is okay.

As the DCG aims to support legacy dca which why I mentioned the mapping documentation you should use the list.label.fields instead of the sorting.headerFields. It was correctly done in the FormatModelLabelSubscriber but not in the list view.

**Comment by amenk** [comment]
@dmolineus I fixed by DCgeneral PR. Maybe you can have a look: https://github.com/contao-community-alliance/dc-general/pull/138

**Comment by dmolineus** [comment]
contao-community-alliance/dc-general#138 looks good so far.

**Comment by amenk** [comment]
great - and what about this PR here?

**Comment by amenk** [comment]
Irgendwas funktioniert noch nicht richtig. Ich bekomme nach composer update wieder die "Meta" Spalten id / pid / usw. angezeigt obwohl die nicht im InputScreen drin sind.

Schaue mir das an. Vll habe ich was im PR vergessen.

**Comment by amenk** [comment]
Well, my PR should be complete. Somehow it is not working anymore. Was something with the propertyNames / headerPropertyNames changed recently?

**Comment by dmolineus** [comment]
Guys, I told you **NOT** to use the header properties for this stuff: https://github.com/MetaModels/core/pull/827#issuecomment-117990536 These are made for the parent view. Now the parent view got broken because you use the header fields for both. The columns AND header fields.

If a column field is set and the parent view is used you the parent view does not work anymore.

**Comment by amenk** [comment]
Sorry... forgot that :-(

What do you think about this one:
https://github.com/amenk/MetaModels-core/commit/3412249c44f9021fcf32c2f5faea0b0505689f35

See also the discussion at:
https://github.com/MetaModels/core/pull/841

**Comment by dmolineus** [comment]
amenk@3412249 looks good as it would fix the mixing of header and column fields. 

Even I think that the [if statement](https://github.com/MetaModels/core/pull/841/files#diff-f276c193ac1702154bc1379b86a7ed8bR948) is not required. The [Builder docs](https://github.com/amenk/MetaModels-core/blob/3412249c44f9021fcf32c2f5faea0b0505689f35/src/MetaModels/DcGeneral/Dca/Builder/Builder.php#L121) does not say anything that a null value can be used. Or do I miss here something, @discordier ? I know there are some conditions to check it in this class but even more there it is not checked.

As the label formatter is the right mapping for the columns, I vote to reopen and merge PR #841 and reject [dc-general#153](https://github.com/contao-community-alliance/dc-general/pull/153).

BTW: There could be setups there the issue does not appear, mainly when the columns in the parent and the child has identical names.

**Comment by discordier** [comment]
@dmolineus can we discuss this in mumble today?

**Comment by discordier** [comment]
Merged #855 in ba9fb8f03396a1ccdf97681b8b030d9a2a3cef00
